presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirection.java

```diff
@@ -145,10 +143,23 @@ public Result apply(TableScanNode scanNode, Captures captures, Context context)
                 // symbol should be mapped in redirected table scan
                 return symbol;
             }
-            // Column pruning after PPD into table scan can remove assignments for filter columns from the scan node
+
+            // validate that redirected types match source types
             Type domainType = requiredFilter.getDomains().get().get(destinationColumn).getType();
-            symbol = context.getSymbolAllocator().newSymbol(destinationColumn, domainType);
             ColumnHandle destinationColumnHandle = destinationColumnHandles.get(destinationColumn);
+            Type redirectedType = metadata.getColumnMetadata(context.getSession(), destinationTableHandle.get(), destinationColumnHandle).getType();
+            if (!domainType.equals(redirectedType)) {
+                throwTypeMismatchException(
+                        destinationTable,
+                        destinationColumn,
+                        redirectedType,
+                        scanNode.getTable(),
+                        sourceColumnHandle,
+                        domainType);
+            }
+
+            // Column pruning after PPD into table scan can remove assignments for filter columns from the scan node
```

Comment:
PPD is a wrong term here. "pushing predicates" 

again, is this a bug/problem?

Reply:
I don't think it's a problem. If redirection runs after a predicate is pushed down into table scan and some of the columns from the filter get pruned from table scan assignments due to column pruning rules, then the redirection rule has to do extra work to "undo" those operations by producing (Project + Filter) nodes and allow the destination connector to handle the projection and predicate pushdowns as supported by it.

Reply:
Updated comment.

This is expected behavior. Once predicate is fully pushed into table scan, assignments for symbols that were part of predicate might potentially be pruned.